### PR TITLE
refactor: refactor modal based components - FE-3828

### LINKED ISSUE #3937: Document Dialog size prop
### Current behaviour
The `size` prop is not shown on the Dialog/Modal/Confirm prop table in storybook.

### Desired behaviour
Add the `size` prop to Dialog/Modal/Confirm prop table in storybook.

https://github.com/Sage/carbon/blob/9abf0aacbf2a4153c0885d81f0a3df3a25f39ce3/src/components/dialog/dialog.style.js#L8-L17

### Suggested solution
N/A

### Describe alternatives you've considered
N/A

### Additional context
Requested by several teams.

### Your environment

| Software         | Version(s) |
| ---------------- | ---------- |
| carbon-react     |68.25.2

### PR DESCRIPTION
**Pull request by mkrds**
BREAKING CHANGE: DialogFullScreen, Dialog, Alert and Confirm components
no longer support enabling background UI interactions or disabling focus trap

### Proposed behaviour
This PR refactors Modal, Dialog, DialogFullScreen, Sidebar, Confirm and Alert components.
This PR also removes support for enabling background UI interactions or disabling focus trap on Dialog, DialogFullScreen, Confirm and Alert components

#Fixes #3937 
### Checklist

- [x] Commits follow our style guide
- [x] Screenshots are included in the PR if useful
- [x] All themes are supported if required
- [x] Unit tests added or updated if required
- [x] Cypress automation tests added or updated if required
- [x] Storybook added or updated if required
- [x] Typescript `d.ts` file added or updated if required
- [x] Carbon implementation and Design System documentation are congruent

### Testing instructions
Full regression on Dialog, DialogFullScreen, Sidebar, Confirm and Alert components

### CODE REVIEW

File: src/components/dialog-full-screen/dialog-full-screen.spec.js [hunk 1]
```diff
@@ -124,7 +111,46 @@ describe("DialogFullScreen", () => {
...
+      expect(window.document.documentElement.style.overflow).toBe("auto");
+      wrapper.setProps({ open: true });
+      expect(window.document.documentElement.style.overflow).toBe("hidden");
+      wrapper.unmount();
+      expect(window.document.documentElement.style.overflow).toBe("auto");
+    });
+  });
+
   describe("dialogTitle", () => {
+    // const headingHeight = 100;
```
Comment (edleeks87) [inline src/components/dialog-full-screen/dialog-full-screen.spec.js:132]: Commented code

File: src/components/dialog-full-screen/dialog-full-screen.spec.js [hunk 2]
```diff
@@ -152,42 +178,41 @@ describe("DialogFullScreen", () => {
...
-
-      it("and the ref to content exists and contains html", () => {
-        expect(wrapper.instance().contentRef.current).toMatchSnapshot();
-      });
-
-      it("and the ref to heading exists", () => {
-        expect(wrapper.instance().headingRef.current).toMatchSnapshot();
-      });
-    });
+    // it("applies proper height to content style depending on heading height", () => {
```
Comment (edleeks87) [inline src/components/dialog-full-screen/dialog-full-screen.spec.js:188]: Commented code

File: src/components/alert/alert.component.js [hunk 3]
```diff
@@ -1,24 +1,43 @@
...
+  /** Allows developers to specify a specific height for the dialog. */
+  height: PropTypes.string,
+  /** Title displayed at top of Alert */
+  title: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
+  /** Subtitle displayed at top of Alert */
+  subtitle: PropTypes.string,
+  /** Size of Alert */
+  size: PropTypes.string,
+  /** Determines if the close icon is shown */
+  showCloseIcon: PropTypes.bool,
```
Comment (edleeks87) [inline src/components/alert/alert.component.js:32]: As we're already making a `BREAKING CHANGE`, I wonder whether we should remove `showCloseIcon` as the same functionality seems to be covered via whether a callback is passed to `onCancel`

File: src/components/confirm/confirm.component.js [hunk 4]
```diff
@@ -108,7 +105,28 @@ Confirm.defaultProps = {
...
+  height: PropTypes.string,
+  /** Title displayed at top of component */
+  title: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
+  /** Subtitle displayed at top of component */
+  subtitle: PropTypes.string,
+  /** Size of component */
+  size: PropTypes.string,
+  /** Determines if the close icon is shown */
+  showCloseIcon: PropTypes.bool,
+  /* Function or reference to first element to focus */
```
Comment (edleeks87) [inline src/components/confirm/confirm.component.js:126]:
```suggestion
  /** Function or reference to first element to focus */
```

File: src/components/confirm/confirm.component.js [hunk 5]
```diff
@@ -108,7 +105,28 @@ Confirm.defaultProps = {
...
+  focusFirstElement: PropTypes.func,
+  /* Disables auto focus functionality on child elements */
```
Comment (edleeks87) [inline src/components/confirm/confirm.component.js:128]:
```suggestion
  /** Disables auto focus functionality on child elements */
```

File: src/components/confirm/confirm.component.js [hunk 6]
```diff
@@ -108,7 +105,28 @@ Confirm.defaultProps = {
...
+  disableAutoFocus: PropTypes.bool,
   // ** A custom event handler when a confirmation takes place */
```
Comment (edleeks87) [inline src/components/confirm/confirm.component.js:130]:
```suggestion
  /** A custom event handler when a confirmation takes place */
```

File: src/components/confirm/confirm.component.js [hunk 7]
```diff
@@ -108,7 +105,28 @@ Confirm.defaultProps = {
...
   onConfirm: PropTypes.func.isRequired,
   // ** Customise the confirm button label */
```
Comment (edleeks87) [inline src/components/confirm/confirm.component.js:132]:
```suggestion
  /** Customise the confirm button label */
```

File: src/components/dialog-full-screen/dialog-full-screen.component.js [hunk 8]
```diff
@@ -1,139 +1,97 @@
...
-  }
-}
+    </Modal>
+  );
+};
 
 DialogFullScreen.defaultProps = {
   open: false,
-  enableBackgroundUI: true,
   showCloseIcon: true,
```
Comment (edleeks87) [inline src/components/dialog-full-screen/dialog-full-screen.component.js:85]: See comment about removing `showCloseIcon`, I don't think defaulting it to `true` does anything as they would still have to pass an `onCancel` callback

File: src/components/dialog/dialog.component.js [hunk 9]
```diff
@@ -83,199 +87,175 @@ class Dialog extends Modal {
...
-          this._dialog = dialog;
-        }}
-        {...dialogProps}
-        {...tagComponent("dialog", {
-          "data-element": "dialog",
-          "data-role": this.props["data-role"],
-        })}
-        onBlur={this.onDialogBlur}
+      <DialogTitleStyle
+        showCloseIcon={showCloseIcon}
```
Comment (edleeks87) [inline src/components/dialog/dialog.component.js:164]: If we remove `showCloseIcon` this could be 
```
        showCloseIcon={!!onCancel}
```

**Review by nicktitchmarsh — DISMISSED** [review]
I think as long as Eds comments are addressed Im happy with this

**Comment by mkrds** [comment]
@edleeks87 It turns out that `showCloseIcon` is actually used in `Confirm` component, so that means it has to be available in `Dialog` as well, that's why I decided to leave it as it is

**Review by edleeks87 — DISMISSED** [review]

**Review by nicktitchmarsh — DISMISSED** [review]

**Comment by harpalsingh** [comment]
There is some alignment with the X we need to look at further, but the paddings in the headers seem to need a review anyway, so we can addrss that seperately together.

File: src/components/dialog/dialog.component.js [hunk 10]
```diff
@@ -222,18 +179,24 @@ const Dialog = ({
...
-              {children}
+              {React.Children.map(children, (child) => {
+                if (child?.type === Form) {
+                  return React.cloneElement(child, {
+                    dialogRef,
+                  });
+                }
+                return child;
+              })}
+              {/* {children} */}
```
Comment (edleeks87) [inline src/components/dialog/dialog.component.js:199]: commented code

File: cypress/features/regression/dialog.feature [hunk 11]
```diff
@@ -2,24 +2,15 @@ Feature: Dialog component
...
-  @negative
-  Scenario Outline: Set out of scope characters to <height>
-    When I open default "Dialog Test" component in noIFrame with "dialog" json from "commonComponents" using "<nameOfObject>" object name
-    Then Dialog height is not set to <height>
-    Examples:
-      | height                       | nameOfObject           |
-      | -1                           | height-1               |
-      | -10                          | height-10              |
-      | !@#$%^*()_+-=~[];:.,?{}&"'<> | heightSpecialCharacter |
+      | 1000   | height1000px |
```
Comment (DlgSHi) [inline cypress/features/regression/dialog.feature:13]:
```suggestion
      | 1000   | height1000 |
```

File: cypress/fixtures/commonComponents/dialog.json [hunk 12]
```diff
@@ -1,21 +1,15 @@
...
-  "height-1": {
-    "height": -1
-  },
-  "height-10": {
-    "height": -10
-  },
-  "heightSpecialCharacter": {
-    "height": "!@#$%^*()_+-=~[];:.,?{}&\"'<>"
+  "height1000px": {
+    "height": "1000px"
```
Comment (DlgSHi) [inline cypress/fixtures/commonComponents/dialog.json]:
```suggestion
    "height": "1000"
```

File: cypress/fixtures/commonComponents/dialog.json [hunk 13]
```diff
@@ -1,21 +1,15 @@
...
   },
-  "height-1": {
-    "height": -1
-  },
-  "height-10": {
-    "height": -10
-  },
-  "heightSpecialCharacter": {
-    "height": "!@#$%^*()_+-=~[];:.,?{}&\"'<>"
+  "height1000px": {
```
Comment (DlgSHi) [inline cypress/fixtures/commonComponents/dialog.json:11]:
```suggestion
  "height1000": {
```

File: cypress/support/step-definitions/alert-steps.js [hunk 14]
```diff
@@ -1,9 +1,18 @@
...
+    resultHeight = viewportHeight - 20;
+  } else {
+    resultHeight = height;
+  }
+
   alertDialogPreview()
-    .should("have.attr", "height")
-    .should("contain", `${height}`);
+    .should("have.css", "height")
+    .should("contain", `${resultHeight}`);
```
Comment (DlgSHi) [inline cypress/support/step-definitions/alert-steps.js]:
```suggestion
    .and("contain", `${resultHeight}px`);
```

**Review by edleeks87 — DISMISSED** [review]

**Review by nicktitchmarsh — DISMISSED** [review]

**Review by DlgSHi — APPROVED** [review]

**Review by edleeks87 — DISMISSED** [review]

**Review by DlgSHi — APPROVED** [review]

**Review by edleeks87 — DISMISSED** [review]

**Review by DlgSHi — APPROVED** [review]

**Review by edleeks87 — APPROVED** [review]

**Review by nicktitchmarsh — APPROVED** [review]

**Pull request merged**